### PR TITLE
Adding in customProps utility function

### PR DESCRIPTION
**Pull request by mjhenkes**
### Summary
Adding a simple utility for separating custom props from expected props.

### Additional Details

Thanks for contributing to Terra. 
@cerner/terra

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

### CODE REVIEW

File: stories/index.jsx [hunk 1]
```diff
@@ -1,9 +1,8 @@
 /* eslint import/no-extraneous-dependencies: ["error", {"devDependencies": true}] */
-import React from 'react';
-import { storiesOf, action, linkTo } from '@kadira/storybook';
-import Utility from '../src/Utility';
+// import React from 'react';
```
Comment (ryanthemanuel) [inline stories/index.jsx]: Should we go ahead and remove this commented code?

Comment (mjhenkes) [inline stories/index.jsx]: I left it incase someone else wanted to add stuff. I don't really have a strong opinion.

Comment (ryanthemanuel) [inline stories/index.jsx]: I think there's enough examples that people could use to recreate the file if necessary.

**Comment by alex-bezek** [comment]
+1

**Comment by mhemesath** [comment]
What is the benefit of doing this over using [destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)

```js
var a, b, customProps;
({a, b, ...customProps} = {a: 10, b: 20, c: 30, d: 40});
```

To use this I believe we need to use a babel plugin to enable it.

**Comment by mhemesath** [comment]
Another good example of using destructuring in react http://reactpatterns.com/#destructuring-arguments

**Comment by mjhenkes** [comment]
I wasn't sure we wanted to pull in another plug in. I'll investigate further if we want to move in that direction.

**Comment by mhemesath** [comment]
Using rest parameter syntax will also make it much more explicit which properties are first class on the component and which are additional custom

**Comment by bjankord** [comment]
I like the idea of using the rest parameter syntax, this section from reactpatterns.com has a nice write-up on how this is used via jsx spread attributes: http://reactpatterns.com/#jsx-spread-attributes.

**Comment by mjhenkes** [comment]
Given
```javascript
const tableCards = (
  <TerraTableCard
    element={<Icon />}
    isTruncated
    key="123"
    className="react-class"
  />
```
Using custom Props
```javascript
 import { customProps } from 'terra-utility';

 const ReactElement = (props) => {

 const extraProps = customProps(props, ReactElement.propTypes);
 // extra props will equal {classname:'react-class', key:'123'}

 return (
    <div {...extraProps}>
      {props.element}
    </div>
  );

 };

 ReactElement.propTypes = {
  element: element,
  isTruncated: PropTypes.bool,
};
```

Using deconstructor and the 'rest parameter '(...).
The rest parameter requires the `babel-plugin-transform-object-rest-spread` plugin to trans-pile.
```javascript
 const ReactElement = ({ element, isTruncated, ...extraProps }) => {
 // extra props will equal {classname:'react-class', key:'123'}

 return (
    <div {...extraProps}>
      {props.element}
    </div>
  );

 };

 ReactElement.propTypes = {
  element: element,
  isTruncated: PropTypes.bool,
};
```
I like the deconstructor and rest parameter syntax and the fact that it eliminates my code. The drawback is that you must list out every declared prop again which feels redundant and a bit annoying. I spend a brief amount of time trying to see if we could turn prop types into variables, but I didn't find a answer. As an example of something we've created with more parameters, I've included what slide panel would look like with this approach. 

Slide Panel example
```javascript
const SlidePanel = ({
  mainContent,
  panelContent,
  panelBehavior,
  panelPosition,
  panelSize,
  isFullscreen,
  isOpen,
  fill,
  ...extraProps
  }) => {
  stuff...
};
```

**Comment by mhemesath** [comment]
> The drawback is that you must list out every declared prop again which feels redundant and a bit annoying

That is true. The alternative I've seen is people bundling all custom HTML attributes into their own explicit property. So

```js
 const ReactElement = (props) => {
 // extra props will equal {classname:'react-class', key:'123'}

 return (
    <div {...props.extraProps}>
      {props.element}
    </div>
  );

 };

 ReactElement.propTypes = {
  element: element,
  isTruncated: PropTypes.bool,
};
```

**Comment by mjhenkes** [comment]
> The alternative I've seen is people bundling all custom HTML attributes into their own explicit property.

This was our approach before with the `attributes` prop. The annoyance here is this:
```javascript
const tableCards = (
  <TerraTableCard
    element={<Icon />}
    isTruncated
    key="123"
    className="react-class"
  />
```
turns into this (approximately, i didn't compile this):
```javascript
const tableCards = (
  <TerraTableCard
    element={<Icon />}
    isTruncated
    extraProps={{ key: '123', className: ['react-class',] }}
  />
```

**Comment by mhemesath** [comment]
Why is that annoying?

**Comment by mjhenkes** [comment]
You have to switch styles from jsx to javascript hash/array style which feels more cumbersome to type for me. It's not a deal breaker, I just want to present all sides.

**Comment by mhemesath** [comment]
FWIW - this is exactly how you set style attributes on elements https://facebook.github.io/react/docs/dom-elements.html#style

So the pattern isn't unique this situation.

**Comment by mjhenkes** [comment]
@mhemesath which approach is your preferred? What do others like?

**Comment by mhemesath** [comment]
I'm fine with either rest params or an explicit property for html attributes. Whatever we do should proobably be consistent throughout all components.

**Comment by mjhenkes** [comment]
I prefer the API for the rest params or the helper over the explicit property. Of those two I like that the helper requires less code, but not enough to fight for it. 

Does everyone agree to go with the rest params style? If so I'll close this pr and update the generator to pull in the babel plugin.

**Pull request closed without merging**